packages/react-core/src/components/ChipGroup/ChipGroup.tsx

```diff
@@ -17,112 +19,163 @@ export interface ChipGroupProps extends React.HTMLProps<HTMLDivElement> {
-    return (
-      <ChipGroupContext.Provider value={headingLevel}>
-        <InnerChipGroup {...this.props} isOpen={isOpen} onToggleCollapse={this.toggleCollapse} />
-      </ChipGroupContext.Provider>
+  renderLabel(id: string) {
+    const { categoryName, tooltipPosition } = this.props;
+    const { isTooltipVisible } = this.state;
+    return isTooltipVisible ? (
+      <Tooltip position={tooltipPosition} content={categoryName}>
+        <span tabIndex={0} ref={this.headingRef} className={css(styles.chipGroupLabel)} aria-hidden="true" id={id}>
+          {categoryName}
+        </span>
+      </Tooltip>
+    ) : (
+      <span ref={this.headingRef} className={css(styles.chipGroupLabel)} aria-hidden="true" id={id}>
+        {categoryName}
+      </span>
     );
   }
 
-  renderChipGroup() {
-    const { className } = this.props;
+  render() {
+    const {
+      categoryName,
+      children,
+      className,
+      isClosable,
+      closeBtnAriaLabel,
+      'aria-label': ariaLabel,
+      onClick,
+      numChips,
+      expandedText,
+      collapsedText,
+      ...rest
+    } = this.props;
     const { isOpen } = this.state;
-    return (
-      <ul className={css(styles.chipGroup, className)}>
-        <InnerChipGroup {...this.props} isOpen={isOpen} onToggleCollapse={this.toggleCollapse} />
-      </ul>
-    );
-  }
+    const collapsedTextResult = fillTemplate(collapsedText as string, {
+      remaining: React.Children.count(children) - numChips
+    });
 
-  render() {
-    const { withToolbar, children } = this.props;
     if (React.Children.count(children)) {
-      return withToolbar ? this.renderToolbarGroup() : this.renderChipGroup();
+      const renderChipGroup = (id: string) => (
+        <div className={css(styles.chipGroup, className, categoryName && styles.modifiers.category)}>
+          {categoryName && this.renderLabel(id)}
+          <ul
+            className={css(styles.chipGroupList)}
+            {...(categoryName && { 'aria-labelledby': id })}
+            {...(!categoryName && { 'aria-label': ariaLabel })}
+            // aria-labelledby={id}
+            // aria-label={ariaLabel}
+            role="list"
+            {...rest}
+          >
+            {isOpen ? (
+              <React.Fragment>
+                {React.Children.toArray(children).map((child, i) => (
+                  <li className={css(styles.chipGroupListItem)} key={i}>
+                    {child}
+                  </li>
+                ))}
+              </React.Fragment>
+            ) : (
+              <React.Fragment>
+                {React.Children.toArray(children).map((child, i) => {
+                  if (i < numChips) {
+                    return (
+                      <li className={css(styles.chipGroupListItem)} key={i}>
+                        {child}
+                      </li>
+                    );
+                  }
+                })}
+              </React.Fragment>
+            )}
```

Comment:
This can be simplified quite a bit.
```tsx
{React.Children.toArray(children)
  .filter((_child, i) => isOpen || i < numChips)
  .map((child, i) => (
    <li className={css(styles.chipGroupListItem)} key={i}>
      {child}
    </li>
  ))}
```

Reply:
Doesn't this create double the runtime of algorithm using this approach instead of the current one?

Reply:
I don't think so. Both algorithms iterate through all of their children which is O(n).